src/audio/google/google_rtc_audio_processing.c

```diff
@@ -665,20 +665,63 @@ static inline void execute_aec(struct google_rtc_audio_processing_comp_data *cd)
 	cd->aec_reference_frame_index = 0;
 }
 
-static int google_rtc_audio_processing_process(struct processing_module *mod,
-					       struct input_stream_buffer *input_buffers,
-					       int num_input_buffers,
-					       struct output_stream_buffer *output_buffers,
-					       int num_output_buffers)
+static void source_copy(struct sof_source *src, int frames, int16_t *dst)
+{
+	size_t chan = source_get_channels(src);
+	size_t samples = frames * chan;
+	size_t bytes = samples * sizeof(int16_t);
+	const int16_t *buf, *bufstart, *bufend;
+	int i, c, err;
+	size_t bufsz;
+
+	err = source_get_data(src, bytes, (void *)&buf, (void *)&bufstart, &bufsz);
+	assert(err == 0);
+	bufend = &bufstart[bufsz];
+
+	for (i = 0; i < frames; i++) {
+		for  (c = 0; c < chan; c++) {
+			*dst++ = *buf++;
+			if (buf >= bufend)
+				buf = bufstart;
```

Comment:
we're now using a trick to avoid checking overflow on each iteration: split into two loops - first until the end, then after wrapping

Reply:
The compiler doesn't unroll that for you?  Pretty sure the optimizer has this covered, I'll check.  If not, sure.

Reply:
I checked.  It actually doesn't, though the generated code for that test is just two instructions.  Really this loop is pretty tight, I don't see a lot to improve, and it's not clear to me that the extra icache footprint of manually unrolling it is going to be a win.  Here's what it looks like for me, the test in question is the load/branch at 0xa005821b:

```asm
a00581fb <mod_process+0x2fb>:
                for  (c = 0; c < chan; c++) {
a00581fb:       c9652481a70e                    { loop  a7, a0058229 <mod_process+0x329>; addi  a4, a1, 88 }
                        *dst[c]++ = ref_to_float(*buf++);
a0058201:       d8120e04256e                    { addi  a6, a5, 2; l32i a14, a4, 0 }
a0058207:       6169                            s32i.n  a6, a1, 24
a0058209:       25dd                            ae_l16si.n      a13, a5, 0
        return (1.0f / REF_SCALE) * (float)x;
a005820b:       397483e46c022d14fed01f          { float.s       aed0, a13, 0; addi      a15, a14, 4; nop; nop }
                        *dst[c]++ = ref_to_float(*buf++);
a0058216:       04f9                            s32i.n  a15, a4, 0
        return (1.0f / REF_SCALE) * (float)x;
a0058218:       2a1000                          mul.s   aed0, aed0, aed1
                        *dst[c]++ = ref_to_float(*buf++);
a005821b:       e00e04                          ae_s32.l.i      aed0, a14, 0
                        if (buf >= bufend)
a005821e:       6158                            l32i.n  a5, a1, 24
a0058220:       033527                          bltu    a5, a2, a0058227 <mod_process+0x327>
                                buf = bufstart;
a0058223:       7158                            l32i.n  a5, a1, 28
a0058225:       6159                            s32i.n  a5, a1, 24

a0058227 <mod_process+0x327>:
                for  (c = 0; c < chan; c++) {
a0058227:       444b                            addi.n  a4, a4, 4

a0058229 <mod_process+0x329>:
        for (i = 0; i < frames; i++) {
a0058229:       330b                            addi.n  a3, a3, -1
a005822b:       738c                            beqz.n  a3, a0058236 <mod_process+0x336>

a005822d <mod_process+0x32d>:
                        *dst[c]++ = ref_to_float(*buf++);
a005822d:       ab7fc561a73e                    { bnez.w15      a7, a00581fb <mod_process+0x2fb>; l32i.n        a5, a1, 24 }
a0058233:       fffc86                          j       a0058229 <mod_process+0x329>
```

Reply:
> The compiler doesn't unroll that for you? Pretty sure the optimizer has this covered, I'll check. If not, sure.

if it did, I'd be greatly impressed! I mean this kind of unwrapping https://github.com/thesofproject/sof/blob/240cc9786ae245ca97fb9e9f4537cfaaab92d6a6/src/audio/eq_fir/eq_fir_generic.c#L38-L57 - the external loop actually runs at most twice, importantly the actual inner sample calculation loop doesn't need to check for wrapping on each new sample.

Reply:
> the test in question is the load/branch at 0xa005821e

just to make this precise, I think it's actually 0xa0058220:
```
a0058220:       033527                          bltu    a5, a2, a0058227 <mod_process+0x327>
```